### DISCUSSION TITLE
Unexpected and unnessery populate issue

### DISCUSSION THREAD

**Discussion by sumitpaultechnoexponent** [mikro-orm/mikro-orm#4111] [Q&A]
**My Mikro orm config**
``` typescript
const config: Options = {
    entities: ['./dist/**/*.entity.js'],
    entitiesTs: ['./src/**/*.entity.ts'],
    baseDir: process.cwd(),
    type: configService.get("DB_TYPE"),
    dbName: configService.get("DB_NAME"),
    user: configService.get("DB_USER"),
    password: configService.get("DB_PASSWORD"),
    host: configService.get("DB_HOST"),
    port: parseInt(configService.get("DB_PORT")),
    metadataProvider: TsMorphMetadataProvider,
    debug: true,
    tsNode: false,
}
```
**My user entity**
``` typescript
@Entity({ customRepository: () => UserRepository })
export class User {

    [EntityRepositoryType]?: UserRepository;

    @PrimaryKey()
    id!: number;

    @Property()
    name!: string;

    @Property()
    email! : string;

    @Property()
    password! : string;
    
    @Enum({
        items : ["APP_USER", "ORG_USER", "SERVICE_PROVIDER_USER", "SERVICE_RECEIVER_USER"],
        default : "SERVICE_RECEIVER_USER"
    })
    type! : string

    @Enum({
        items : ["SUPERADMIN", "ADMIN", "MANAGER", "SUPPORT", "OWNER", "MEMBER", "SALES", "GP", "CUSTOMER"],
        default : "CUSTOMER"
    })
    role! : string

    @Property()
    createdAt = new Date();

    @Property({
        onUpdate: () => new Date()
    })
    updatedAt = new Date();

    @OneToMany(() => Shop, shop => shop.owner)
    shop = new Collection<Shop>(this);

    @OneToMany(() => Product, product => product.owner)
    product = new Collection<Product>(this);

    @OneToMany(() => Permission, permission => permission.user)
    permission = new Collection<Permission>(this);
    

    constructor( name : string, email : string, password: string, type : string, role : string ) {
        this.name = name;
        this.email = email;
        this.password = password;
        this.type = type;
        this.role = role;
    }

}
```

**My Shop Entity**

``` typescript
@Entity({ customRepository: () => ShopRepository })
export class Shop {

    [EntityRepositoryType]?: ShopRepository;

    @PrimaryKey()
    id!: number;

    @Property()
    name!: string;

    @Property()
    description?: string;

    @Property()
    createdAt = new Date();

    @Property({
        onUpdate: () => new Date()
    })
    updatedAt = new Date();

    @OneToMany(() => Product, product => product.shop)
    products = new Collection<Product>(this);

    @ManyToOne()
    owner!: User;

    constructor(name: string, description: string, owner: User) {
        this.name = name;
        this.description = description;
        this.owner = owner;
    }

}
```

**My Product Entity**

``` typescript
@Entity({ customRepository: () => ProductRepository })
export class Product {

    [EntityRepositoryType]?: ProductRepository;

    @PrimaryKey()
    id!: number;

    @Property()
    name!: string;

    @Property()
    description?: string;

    @Property()
    amount! : number

    @ManyToOne()
    shop!: Shop;

    @Property()
    createdAt = new Date();

    @Property({
        onUpdate: () => new Date()
    })
    updatedAt = new Date();
    
    @ManyToOne()
    owner!: User;

    constructor(name: string, description: string, amount: number, shop: Shop, owner : User) {
        this.name = name;
        this.description = description;
        this.amount = amount;
        this.shop = shop;
        this.owner = owner;
    }

}
```

**My query**
``` typescript
const shops = this.shopRepository.findAll({
            orderBy:{id:'asc'},
            populate: ['products', 'owner']          
        });
``` 

**Expected Result**
``` json
[
    {
        "id": 1,
        "name": "string",
        "description": "string",
        "createdAt": "2023-03-09T08:22:54.000Z",
        "updatedAt": "2023-03-09T08:22:54.000Z",
        "products": [
            {
                "id": 1,
                "name": "string",
                "description": "string",
                "amount": 10,
                "createdAt": "2023-03-09T08:23:28.000Z",
                "updatedAt": "2023-03-09T08:23:28.000Z",
                "shop": 1,
                "owner": 1
            }
        ],
        "owner": {
            "id": 1,
            "name": "Sumit1",
            "email": "mailto:sumit1@yopmail.com",
            "password": "$2a$10$K92TcXRwjMH7ol/B.yLlXuiVaQwikHUyS8B.CUUj6oyQqTBsmldgm",
            "type": "APP_USER",
            "role": "SUPERADMIN",
            "createdAt": "2023-03-09T08:19:09.000Z",
            "updatedAt": "2023-03-09T08:19:09.000Z"
        }
    }
]
```

**Actual Result**
``` json
[
    {
        "id": 1,
        "name": "string",
        "description": "string",
        "createdAt": "2023-03-09T08:22:54.000Z",
        "updatedAt": "2023-03-09T08:22:54.000Z",
        "products": [
            {
                "id": 1,
                "name": "string",
                "description": "string",
                "amount": 10,
                "createdAt": "2023-03-09T08:23:28.000Z",
                "updatedAt": "2023-03-09T08:23:28.000Z",
                "shop": 1,
                "owner": {
                    "id": 1,
                    "name": "Sumit1",
                    "email": "mailto:sumit1@yopmail.com",
                    "password": "$2a$10$K92TcXRwjMH7ol/B.yLlXuiVaQwikHUyS8B.CUUj6oyQqTBsmldgm",
                    "type": "APP_USER",
                    "role": "SUPERADMIN",
                    "createdAt": "2023-03-09T08:19:09.000Z",
                    "updatedAt": "2023-03-09T08:19:09.000Z"
                }
            }
        ],
        "owner": {
            "id": 1,
            "name": "Sumit1",
            "email": "mailto:sumit1@yopmail.com",
            "password": "$2a$10$K92TcXRwjMH7ol/B.yLlXuiVaQwikHUyS8B.CUUj6oyQqTBsmldgm",
            "type": "APP_USER",
            "role": "SUPERADMIN",
            "createdAt": "2023-03-09T08:19:09.000Z",
            "updatedAt": "2023-03-09T08:19:09.000Z"
        }
    }
]
```

Now my question is I want to populate owner only under the shop not under the product.
And am not able to do this, Can you please suggest something to achieve my expected result.
Thank you

**Comment by B4nan** [comment]
This should work out of box if you use populate hints. Are you on latest version? Can you turn this into executable repro, ideally as a single script?

Maybe you load that problematic owner in another way too, e.g. in some middleware?

**Reply by sumitpaultechnoexponent** [reply]
Thanks for your help. Actually am not getting this point 'executable repro' .
Am sharing my public repo and upload everything in this repo
Link : https://github.com/PrimeDominus/nestjs-mikro
Can please check this one time and suggest me what I need to do to achieve my goal.
Thank you

**Reply by B4nan** [reply]
I am asking for [a minimal but complete reproduction](https://stackoverflow.com/help/minimal-reproducible-example). Reproduction should be as simple as possible while still showing the problem, you provided a rather huge app, with a lot of unrelated stuff - dozens of files, while I wanted to have just one, as short as possible. 

I'll be happy to help, but I need to see something much simpler so I can quickly navigate it. I can't help you debug your apps, not for free.

Also, I told you to upgrade to the latest version, and you gave me a repro that is using v5.2.2? :]

**Reply by sumitpaultechnoexponent** [reply]
I have removed all unnecessary code, and update the repo with minimal setup.
You just need to spend 2mins to see my error and also I have update mikro orm latest version.
Link : https://github.com/PrimeDominus/nestjs-mikro
Please let me know if you need anything else.
Thank you for helping me.

**Reply by B4nan** [reply]
Not really, this is far away from minimal. I am asking for a script, ideally a single file. You gave me a webapp (so the opposite of a script), with many files, looks like ~10 files for a single entity. I don't want to browse that code to see what is broken and where. 

I mean I will be happy to look into this, but it won't be 2 minutes, I will spend more just by cloning and installing it. And debugging it will still require me to come up with an actual minimal reproduction - a test case. So if you want me to look into this rather sooner than later, please simplify this more - remove the whole nestjs layer. You will probably see the problem is that part (so not in the ORM itself), and that is where you can point me to a specific place. This is the part you can debug as well, helping me to focus on things you cannot do yourself.

**Reply by sumitpaultechnoexponent** [reply]
Also I made thie repo more minimal.
Link : https://github.com/PrimeDominus/nestjs-mikro

1. clone this repo
2. npm install
3. npm run start:dev1
4. npx mikro-orm seeder:run --class=MainSeeder
5. npm run start:dev1

Then login with below credenials [/v1/auth/login  -> POST]
``` json
{
  "email": "sp-1@yopmail.com",
  "password": "1234"
}
```
then run v1/shop (GET) route with login token,  then you will see the response

Thank you

**Reply by sumitpaultechnoexponent** [reply]
> Not really, this is far away from minimal. I am asking for a script, ideally a single file. You gave me a webapp (so the opposite of a script), with many files, looks like ~10 files for a single entity. I don't want to browse that code to see what is broken and where.
> 
> I mean I will be happy to look into this, but it won't be 2 minutes, I will spend more just by cloning and installing it. And debugging it will still require me to come up with an actual minimal reproduction - a test case. So if you want me to look into this rather sooner than later, please simplify this more - remove the whole nestjs layer. You will probably see the problem is that part (so not in the ORM itself), and that is where you can point me to a specific place. This is the part you can debug as well, helping me to focus on things you cannot do yourself.

If you need without nestjs I will provide you the setup shortly.
Thank you

**Comment by sumitpaultechnoexponent** [comment]
Hello I have reduce all extra code from my repo.
Please check if you need anything else 
Link : https://github.com/PrimeDominus/nestjs-mikro

Thank you for your help

**Reply by sumitpaultechnoexponent** [reply]
any update on this?

**Comment by sumitpaultechnoexponent** [comment]
hello do you need my script like this ?

``` javascript
import { Collection, Entity, ManyToOne, OneToMany, PrimaryKey, Property } from '@mikro-orm/core';
import { MikroORM } from '@mikro-orm/postgresql';

@Entity()
class User {

    @PrimaryKey()
    id!: number;

    @Property()
    name!: string;

    @Property()
    email!: string;

    @OneToMany(() => Shop, shop => shop.owner)
    shop = new Collection<Shop>(this);

    @OneToMany(() => Product, product => product.owner)
    product = new Collection<Product>(this);


    constructor(name: string, email: string) {
        this.name = name;
        this.email = email;
    }

}

@Entity()
class Shop {

    @PrimaryKey()
    id!: number;

    @Property()
    name!: string;

    @OneToMany(() => Product, product => product.shop)
    products = new Collection<Product>(this);

    @ManyToOne()
    owner!: User;

    constructor(name: string, owner: User) {
        this.name = name;
        this.owner = owner;
    }

}

@Entity()
export class Product {

    @PrimaryKey()
    id!: number;

    @Property()
    name!: string;

    @ManyToOne()
    shop!: Shop;

    @ManyToOne()
    owner!: User;

    constructor(name: string, shop: Shop, owner: User) {
        this.name = name;
        this.shop = shop;
        this.owner = owner;
    }

}

let orm: MikroORM;

beforeAll(async () => {
    orm = await MikroORM.init({
        entities: [User, Shop, Product],
        type: "postgresql",
        dbName: "mikroORMNest",
        user: "root",
        password: "root",
        host: "localhost",
        port: 5432
    });
    await orm.schema.refreshDatabase();
});

beforeEach(async () => {
    await orm.schema.clearDatabase();

    await orm.em.create(User, {
        "name": "s1",
        "email": "sp-1@yopmail.com"
    });
    await orm.em.create(User, {
        "name": "sp-2",
        "email": "sp-2@yopmail.com"
    });
    await orm.em.create(Shop, {
        "name": "shop-1",
        "owner": 1
    });
    await orm.em.create(Product, {
        "name": "product-1",
        "shop": 1,
        "owner": 1
    });
    await orm.em.create(Product, {
        "name": "product-2",
        "shop": 1,
        "owner": 2
    });

    await orm.em.flush();
    orm.em.clear();
});


test('test with populate', async () => {
    const shop = orm.em.findOne(Shop, {}, {
        populate: ['products', 'owner']
    });

    console.log(shop);
    // here shop should include owner but products should not include product
    // but the products also include owner (only those owner matching with shop owner)
    
});


```

**Reply by B4nan** [reply]
Yes, thanks, that's helpful. 

Few notes about the repro, I guess you haven't really tried to run it, right?
- it fails on validation error, as you can't use `findOne` with empty condition
- it is missing an await on the very same line
- you are awaiting `em.create` which is a sync method
- and lastly, logging the entity instance is irrelevant, you need to serialize the value, runtime entity shape should have that owner entity available, that is how the identity map works - you can't have the same entity in two states

But after fixing those small problems, I can see it reproduces the problem. 

I will try to look into this soon, but it won't be sooner than next week probably.

**Reply by B4nan** [reply]
Created this issue so I dont forget https://github.com/mikro-orm/mikro-orm/issues/4138

**Comment by sumitpaultechnoexponent** [comment]
thank for your reply
Yes I havent run this. I just want to send you my problem as minimal as I can.
Please let me know if you fix this.
